8-writing-advanced-tasks/README.md

```diff
@@ -0,0 +1,228 @@
+# Writing advanced tasks
+
+> **Difficulty**: Intermediate
+
+> **Time**: Approximately 10 minutes
+
+In this exercise you will write a task with metadata. Including a task that supports_noop.
+
+# Prerequisites
+
+For the following exercises you should already have `bolt` installed and have a few nodes (either Windows or Linux) available to run commands against. The following guides will help:
+
+1. [Installing Bolt](../1-installing-bolt)
+1. [Acquiring nodes](../2-acquiring-nodes)
+1. [Writing tasks](../5-writing-tasks)
+
+# Writing your first task with metadata
+
+Writing metadata helps those that consume your tasks know how to provide input to the tasks that you write.  Metadata for a task is written in JSON and is expected to have the same name as your task with a `.json` extension. So if you write a task called `foo_bar.py` you should write a corresponding `foo_bar.json`.  Lets start by writing a simple task that formats the parameters a user gives it.      
+
+Save the following file to `modules/exercise8/tasks/foo_bar.py`:
+```
+#! /usr/bin/env python
+
+"""
+This script prints the values and types passed to it via standard in.  It will
+return a JSON string with a parameters key containing objects that describe
+the parameters passed by the user.
+"""
+
+import json
+import sys
+
+def make_serializable(object):
+  if isinstance(object, unicode):
+    return object.encode('utf-8')
+  else:
+    return object
+
+data = json.load(sys.stdin)
+
+message = "Congratulations on writing your metadata!  Here are the keys and the values that you passed to this task."
+
+result = {'message': message, 'parameters': []}
+for key in data:
+    k = make_serializable(key)
+    v = make_serializable(data[key])
+    t = type(data[key]).__name__
+    param = {'key': k, 'value': v, 'type': t}
+    result['parameters'].append(param)
+
+print(json.dumps(result))
+```
+
+Now lets write the accompanying metadata.  Well specify the parameters as types such as `"type": "Integer"`  which will help us validate the users input as an `Integer`.  Save the following file to `modules/exercise8/tasks/great_metadata.json`:
+```
+{
+  "description": "An exercise in writing great metadata",
+  "input_method": "stdin",
+
+  "parameters": {
+    "name": {
+      "description": "The description for the 'name' parameter",
+      "type": "String"
+    },
+    "recursive": {
+      "description": "The description for the 'recursive' parameter",
+      "type": "Boolean"
+    },
+    "action": {
+      "description": "The description for the 'action' parameter",
+      "type": "Enum[stop, start, restart]"
+    },
+    "timeout": {
+      "description": "The description for the 'timeout' parameter",
+      "type": "Optional[Integer]"
+    }
+  }
+}
+```
+
+# Using your task with metadata
+
+Use the following command to show to test that your task is listed with its description. 
+
+```
+$ bolt task show 
```

Comment:
This needs a --modulepath option

Reply:
In lab 6 we instruct users to write a config file with modulepath configured.  https://github.com/puppetlabs/tasks-hands-on-lab/tree/master/6-downloading-and-running-existing-tasks#clone-the-control-repo-and-configure-bolts-modulepath  Should we build on this instruction or assume they will not do it?